product_stock_state/models/product_template.py

```diff
@@ -0,0 +1,17 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017-Today GRAP (http://www.grap.coop).
+# @author Sylvain LE GAL <https://twitter.com/legalsylvain>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    stock_state_threshold = fields.Float(
```

Comment:
To avoid performance issue, this field should be a stored computed field with an inverse method.... The inverse should put the value into an other field while the compute should looks first if the manual value is set and if not look into the category where the field should also be computed....
ie

INTO `product.template`

```python
stock_state_threshold = fields.Float(
    compute='_compute_stock_state_threshold',
    inverse='_inverse_stock_state_threshold',
    store=True)
manual_stock_state_threshold = fields.Float()

@api.multi
def  _inverse_stock_state_threshold(self):
    for rec in self:
        rec.manual_stock_state_threshold = self.stock_state_threshold

@api.multi
@api.depends('categ_id.stock_state_threshold', 'manual_stock_state_threshold')
def _compute_manual_stock_state_threshold(self):
    for rec in self:
        rec.stock_state_threshold = rec.manual_stock_state_threshold or rec.categ_id.stock_state_threshold
```

INTO `product.category`

```python
stock_state_threshold = fields.Float(
    compute='_compute_stock_state_threshold',
    inverse='_inverse_stock_state_threshold',
    store=True)
manual_stock_state_threshold = fields.Float()

@api.multi
def  _inverse_stock_state_threshold(self):
    for rec in self:
        rec.manual_stock_state_threshold = self.stock_state_threshold

@api.multi
@api.depends('parent_id.stock_state_threshold', 'manual_stock_state_threshold')
def _compute_manual_stock_state_threshold(self):
    for rec in self:
        rec.stock_state_threshold = rec.manual_stock_state_threshold or rec.parent_id.stock_state_threshold

```

Once it's done the method `_get_stock_state_threshold` can be simplified and we avoid to recursively search into categories since all the values are stored into the database....

```python
 def _get_stock_state_threshold(self):
        self.ensure_one()
        threshold = self.stock_state_threshold
        if not threshold:
            # try to get threshold from current company
            threshold = self.env.user.company_id.stock_state_threshold
        return threshold
```

Reply:
ping @sebastienbeau

Reply:
Hi I will work on it tomorrow.
I do not know if it's the best solution to put a inverse method here.
Indeed if for some reason odoo try to write again the value of "stock_state_threshold" that was computed (onchange, cache...) It will write it in the manual field.
Maybe it's better to show
- the 2 fields in the view when we edit the categories (Threshold, Force Threshold)
- show the computed field in readonly
What do you think?